### PR TITLE
fix gateway merge on same port

### PR DESCRIPTION
**Pull request by ramaraochavali**
https://github.com/istio/istio/issues/33405

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/model/gateway_test.go [hunk 1]
```diff
@@ -109,9 +113,17 @@ func TestMergeGateways(t *testing.T) {
 			"http-http2-server-config",
 			[]config.Config{gwHTTPWildcard, gwHTTP2Wildcard},
 			1,
-			2,
+			1,
```
Comment (ramaraochavali) [inline pilot/pkg/model/gateway_test.go:116]: This is not related to this change. This was never being validated because test case issue.

File: releasenotes/notes/33405.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - 33405
+releaseNotes:
+  - |
+    **Fixed** a bug where multiple gateways on same port with SIMPLE and PASSTHROUGH modes is not working correctly.
```
Comment (shamsher31) [inline releasenotes/notes/33405.yaml:8]:
```suggestion
    **Fixed** a bug where multiple gateways on the same port with `SIMPLE` and `PASSTHROUGH` modes are not working correctly.

```

File: pilot/pkg/model/gateway.go [hunk 3]
```diff
@@ -231,10 +231,11 @@ func MergeGateways(gateways []gatewayWithInstances) *MergedGateway {
 							log.Warnf("TLS server without TLS options %s %s", gatewayName, s.String())
 							continue
 						}
-						if mergedServers[serverPort] == nil {
-							mergedServers[serverPort] = &MergedServers{Servers: []*networking.Server{s}}
-						} else {
-							mergedServers[serverPort].Servers = append(mergedServers[serverPort].Servers, s)
+						for port := range mergedServers {
+							if port.Number == serverPort.Number {
+								mergedServers[port].Servers = append(mergedServers[port].Servers, s)
```
Comment (howardjohn) [inline pilot/pkg/model/gateway.go:236]: I am a bit concerned about this. In buildGatewayListeners, we use the key.Protocol field, which will now be wrong here, as with .Bind. Also, its n^2 and folks have O(10k) gateways so it may be a scalability issue

**Comment by ramaraochavali** [comment]
@howardjohn changed the gateway listener build logic to handle this case. PTAL

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 4]
```diff
@@ -59,7 +64,7 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
 
 	actualWildcard, _ := getActualWildcardAndLocalHost(builder.node)
 	errs := istiomultierror.New()
-	listeners := make([]*listener.Listener, 0)
+	mutableopts := make(map[string]mutableListenerOpts)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:67]: can you indicate what the key is?

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 5]
```diff
@@ -124,54 +129,75 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
...
-						server, mergedGateway.GatewayNameForServer[server])
-					filterChainOpts = append(filterChainOpts, tcpChainOpts...)
-					for i := 0; i < len(tcpChainOpts); i++ {
-						filterChains = append(filterChains, istionetworking.FilterChain{ListenerProtocol: istionetworking.ListenerProtocolTCP})
+					// This is the case of TCP or PASSTHROUGH.
+					// Check if we have already built a listener for this - this happens when there are multiple gateways
+					// one with SIMPLE TLS and another with PASSTHROUGH. In that case we should append the filter chains
+					// to the existing listener.
+					mops, exists := mutableopts[lname]
+					if exists {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:165]: This seems a bit odd that only in this spot we are messing with other listeners, and everywhere else we just create a big list of filter chains. I  am a bit concerned this doesn't actually handle all cases. For example, what if we have HTTP and GRPC ports? don't we have the same issue? And why do we have the code here, but not for HTTPS - can't the same happen in that case?

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: Yes. Agree that it is bit odd that is handled in this case. I was n't able to wrap my head around all possibilities. But thinking about it a bit more, we can make this generic. I have changed. I think we will have to add more tests in future.

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: I think te allowed setup are:
* N HTTP servers (megred into a single filtre chain)
* N TLS/HTTPS servers and 0 or 1 TCP server
* 1 TCP server
Is that accurate?

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: Yes. It seems correct.

**Comment by ramaraochavali** [comment]
@howardjohn added more test cases to cover scenarios. Seems to work fine. Let me know if you would like any more cases to be added.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 6]
```diff
@@ -114,64 +120,69 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
...
+			mutable = &MutableListener{
+				MutableObjects: istionetworking.MutableObjects{
+					// Note: buildListener creates filter chains but does not populate the filters in the chain; that's what
+					// this is for.
+					FilterChains: newFilterChains,
+				},
+			}
+			mutableopts[lname] = mutableListenerOpts{mutable: mutable, opts: opts}
+		} else {
+			mopts.opts.filterChainOpts = append(mopts.opts.filterChainOpts, opts.filterChainOpts...)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:183]: There is a small bug here. It definitely exists in current master, not sure if it did in 1.9. The iteration of MergedServers is randomized order due to golang maps, so we may get different filter chains. This doesn't matter to envoy logic, but *might* trigger incorrect behavior in some places like envoy diffing, istioctl diffing, istiod caching (not a concern now or in the near future, but in theory...)

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 7]
```diff
@@ -184,18 +195,21 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
...
 		}
-
+	}
+	listeners := make([]*listener.Listener, 0)
+	for _, ml := range mutableopts {
+		ml.mutable.Listener = buildListener(*ml.opts, core.TrafficDirection_OUTBOUND)
 		// Filters are serialized one time into an opaque struct once we have the complete list.
-		if err := mutable.build(opts); err != nil {
-			errs = multierror.Append(errs, fmt.Errorf("gateway omitting listener %q due to: %v", mutable.Listener.Name, err.Error()))
+		if err := ml.mutable.build(*ml.opts); err != nil {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:203]: nit: can we just make `ml.opts` not a pointer?

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/gateway.go:203]: If not no problem

**Review by howardjohn — APPROVED** [review]
Looks good, I like the new changes :+1:
